### PR TITLE
Jed/python packaging

### PR DESCRIPTION
**Pull request by jedbrown**
This tentatively adds `pip install libceed` using (binary) wheels. I have it currently on the testing server, so you can try
```
pip install -i https://test.pypi.org/simple/ libceed
```
If this (finally) works, I'd prefer for the Jupyter notebooks #485 to use it instead of going through the tedium of installing from source. If this passes, I'll move the release to main pypi.org.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -26,64 +27,84 @@ def version():
...
+        return f.readlines()
+
+class libceed_build_ext(build_ext):
+    def run(self):
+        self.make_libceed_so()
+        build_ext.run(self)
+
+    def make_libceed_so(self):
+        import subprocess
+        subprocess.check_call(['make', '-j', 'MARCHFLAGS=-march=generic', '-B'])
```
Comment (dalcinl) [inline setup.py]: Maybe allow users to pass `MARCHFLAGS` via environment variable? Then you set it to `-march=generic` for the docker wheel build. Installs from source can then take advantage of optimization.

Comment (jedbrown) [reply setup.py]: Thanks; updated.

**Comment by valeriabarra** [comment]
@jedbrown has something on the testing server changed?
I have tried `make -B` and then
```
pip3 install -i https://test.pypi.org/simple/ libceed

Collecting libceed
  Could not find a version that satisfies the requirement libceed (from versions: )
No matching distribution found for libceed
```

**Comment by valeriabarra** [comment]
Once this build is defined, we should also update the READMEs with current instructions

**Comment by jedbrown** [comment]
@valeriabarra I make a virtualenv in `/tmp`. (You don't need libCEED or even compilers on your machine to do this.)
```
(VENV) /tmp/jed$ pip install cffi numpy
Collecting cffi
  Using cached cffi-1.14.0-cp38-cp38-manylinux1_x86_64.whl (409 kB)
Collecting numpy
  Using cached numpy-1.18.2-cp38-cp38-manylinux1_x86_64.whl (20.6 MB)
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Installing collected packages: pycparser, cffi, numpy
Successfully installed cffi-1.14.0 numpy-1.18.2 pycparser-2.20
(VENV) /tmp/jed$ pip install -i https://test.pypi.org/simple/ libceed
Looking in indexes: https://test.pypi.org/simple/
Collecting libceed
  Downloading https://test-files.pythonhosted.org/packages/a3/df/31482289d80fae065d17089db3623f2871ebe3c324176f728adcdaa44c10/libceed-0.5.2-cp38-cp38-manylinux2014_x86_64.whl (237 kB)
     |████████████████████████████████| 237 kB 2.6 MB/s 
Requirement already satisfied: numpy in ./VENV/lib/python3.8/site-packages (from libceed) (1.18.2)
Requirement already satisfied: cffi in ./VENV/lib/python3.8/site-packages (from libceed) (1.14.0)
Requirement already satisfied: pycparser in ./VENV/lib/python3.8/site-packages (from cffi->libceed) (2.20)
Installing collected packages: libceed
Successfully installed libceed-0.5.2
```

The trouble with dependencies will go away once I post to the real pypi.org, but you can't update packages without changing their names/versions so I want to get it right the first time.

**Comment by dalcinl** [comment]
@jedbrown A minor aesthetic suggestion. Right now, the cffi extension module `_ceed_cffi.abi3.so` ends installed outside the `libceed` package directory, and goes to the top-level site-packages directory. I consider this bad packaging practice, though this is what cffi "recomends" in some examples out there. With some minor modifications to import statements here and there, plus prefixing`ffibuilder.set_source("libceed._ceed_cffi", ...` in the builder script, you could cleanup things and fix this admittedly minor annoyance.

**Comment by jedbrown** [comment]
@dalcinl Yes, I don't know why they recommend it that way, but I agree it would be better to place that module under `libceed`.

**Comment by valeriabarra** [comment]
@jedbrown  I might be doing something wrong, of course, because I usually don't use `virtualenv` or `venv`, but I have used `conda` for package/environment management. But now that I have tried to use `venv` and created an environment in my `/tmp/valeria` folder, it still gives me the same error.

```
valeria@work:/tmp/valeria$ python3 -m venv VENV

valeria@work:/tmp/valeria$ ls VENV/
bin  include  lib  lib64  pyvenv.cfg  share

valeria@work:/tmp/valeria$ source VENV/bin/activate

(VENV) valeria@work:/tmp/valeria$ pip3 install cffi numpy
Collecting cffi
  Using cached https://files.pythonhosted.org/packages/f1/c7/72abda280893609e1ddfff90f8064568bd8bcb2c1770a9d5bb5edb2d1fea/cffi-1.14.0-cp36-cp36m-manylinux1_x86_64.whl
Collecting numpy
  Using cached https://files.pythonhosted.org/packages/07/08/a549ba8b061005bb629b76adc000f3caaaf881028b963c2e18f811c6edc1/numpy-1.18.2-cp36-cp36m-manylinux1_x86_64.whl
Collecting pycparser (from cffi)
  Using cached https://files.pythonhosted.org/packages/ae/e7/d9c3a176ca4b02024debf82342dab36efadfc5776f9c8db077e8f6e71821/pycparser-2.20-py2.py3-none-any.whl
Installing collected packages: pycparser, cffi, numpy
Successfully installed cffi-1.14.0 numpy-1.18.2 pycparser-2.20

(VENV) valeria@work:/tmp/valeria$ pip install -i https://test.pypi.org/simple/ libceed
Collecting libceed
  Could not find a version that satisfies the requirement libceed (from versions: )
No matching distribution found for libceed
```

**Comment by jedbrown** [comment]
@valeriabarra What do you have for `pip --version`?

**Comment by valeriabarra** [comment]
> @valeriabarra What do you have for `pip --version`?

Thanks, now that I have upgraded pip inside the VENV, it has worked just fine.

**Comment by jedbrown** [comment]
I'll step back to `manylinux2010`. I had used `manylinux2014` ('cuz why not newer, right), but I see you need pretty new pip for that.

**Comment by valeriabarra** [comment]
@jedbrown now that I've got my venv setup, I have tested that libCEED's build is fine and I was able to test the jupyter-notebooks involving the python interface (tutorials2-7) and everything seems to work smoothly, so I can update the build instructions on those, but we also have [a jupyter-notebook](https://github.com/CEED/libCEED/blob/valeria/jupyter-notebook-tutorials/examples/tutorials/tutorial1-ceed.ipynb) which involves the libCEED's standalone examples.

**Comment by jedbrown** [comment]
It should work now with the older pip. I'll deploy wheels for real after merging this (and perhaps #485).

**Comment by valeriabarra** [comment]
Thanks @jedbrown , is it ok to merge this?

**Comment by valeriabarra** [comment]
Thanks @jedbrown  for setting up the PyPI libCEED project and the binder.

I'm going to merge this PR and work on #485 if that's ok now.

**Pull request merged**